### ISSUE TITLE
Can not pass the whole pattern to a function.

### ISSUE THREAD

**Issue by xtradev** [BartJongejan/Bracmat#3]
To parse a file (hosts file) I created a function. Here is one its line in question:

```
@(!line:!linePrefix ?host !lineSuffix)
```

I want `lineSuffix` to be expanded to a pattern and be passed as an argument. I call it like so:

```
extract'("mvps.txt","mvps.out",\r\n,1417,"0.0.0.0 ",(" " ?|));
```

the arguments are captured as:

```
!arg:(?inFileName, ?outFileName, ?lineSeparator, ?startOffset, ?linePrefix, ?lineSuffix)
```

And it does not work... It does, if I change the line in question to:

```
@(!line:!linePrefix ?host (!lineSuffix|))
```

and the call to:

```
extract'("mvps.txt","mvps.out",\r\n,1417,"0.0.0.0 ",(" " ?));
```

effectively splitting the pattern and making a part of it a part of the function.

Supposedly, I do not clearly understand where, at what stage, should I postpone the expansion. I use single quote here and have experimented with `=` and `!!`, passed the pattern as a string and tried to escape the `|`. But no success.

**Comment by xtradev** [comment]
Here is the function:

```
( extract
= inFileName, outFileName, lineSeparator, startOffset, linePrefix, lineSuffix, hosts
.
!arg:(?inFileName, ?outFileName, ?lineSeparator, ?startOffset, ?linePrefix, ?lineSuffix)
& fil$(!inFileName,r)
& fil$(,STR,!lineSeparator)
& fil$(,SET,!startOffset)
&   whl
  ' ( fil$:(?line.?)
    & `(@(!line:!linePrefix ?host !lineSuffix)
       & !host \n !hosts:?hosts
       )
    )
& (fil$(,SET,-1)|)
& put$(str$!hosts,!outFileName,NEW)
);
```

**Comment by xtradev** [comment]
```
extract'("mvps.txt","mvps.out",\r\n,1417,  "0.0.0.0 ",$(" " ?|));
```

`$` also does not help (though it is not within a pattern...)

**Comment by BartJongejan** [comment]
The pattern for the line suffix is transferred to the <code>extract</code> function. So far so good. If we now look at the body of the <code>extract</code> function, we have a variable <code>arg</code> that is bound to an unevaluated expression:

```
"mvps.txt","mvps.out",\r\n,1417,"0.0.0.0 ",(" " ?|)
```

You want to keep that expression unevaluated, or at least until you have brought the patterns for the line prefix and suffix in safety. However, when you do <code>!arg</code>, the resulting expression
is evaluated. Worst of all, the subexpression <code>(" " ?|)</code> is evaluated to <code>(" " ?)</code>. 

What you want to do is pattern matching in the unevaluated expression, so you can't use <code>!arg</code>. You need to use macro expansion, like so 

```
'($arg)
```

Use this as the subject of the match operation. <code>'($arg)</code> evaluates to, in this case,

```
="mvps.txt","mvps.out",\r\n,1417,"0.0.0.0 ",(" " ?|)
```

Notice the <code>=</code> operator. The rhs of the <code>=</code> operator is not evaluated and that is what we want.

In the pattern we need to match the <code>=</code> operator as well, so we start with that:

```
=?inFileName, ?outFileName, ?lineSeparator, ?startOffset, ?linePrefix, ?lineSuffix
```

Therefore instead of 

```
!arg:(?inFileName, ?outFileName, ?lineSeparator, ?startOffset, ?linePrefix, ?lineSuffix)
```

you should write

```
'($arg):(=?inFileName, ?outFileName, ?lineSeparator, ?startOffset, ?linePrefix, ?lineSuffix)
```

Another solution, the one I prefer, is to write

```
extract$("mvps.txt","mvps.out",\r\n,1417,(="0.0.0.0 "),(=" " ?|));
```

instead of

```
extract'("mvps.txt","mvps.out",\r\n,1417,"0.0.0.0 ",(" " ?|));
```

and to have this at the start of the body of the extract function:

```
!arg:(?inFileName, ?outFileName, ?lineSeparator, ?startOffset, (=?linePrefix), (=?lineSuffix))
```

The reason is that this clearly indicates what exactly shouldn't be evaluated and allows the rest to be evaluated. The advantage becomes especially clear in a case like the following, where some of the parameters are complex expressions. If you do

```
  mvps:?basename
& extract'(str$(!basename ".txt"),str$(!basename ".out"),\r\n,1417,"0.0.0.0 ",(" " ?|));
```

the expression <code>str$(!basename ".txt")</code> is assigned to <code>inFileName</code> and the expression <code>str$(!basename ".out")</code> is assigned to <code>outFileName</code>. If you then do <code>!inFileName</code> in the body of <code>extract</code>, the expression <code>str$(!basename ".txt")</code> is evaluated. Now, if <code>basename</code> happens to be a locally declared variable in the <code>extract</code> function, you probably get the wrong result. (In your case, where <code>basename</code> isn't a local variable in the <code>extract</code> function and <code>!inFileName</code> isn't evaluated many times, it works fine.)

On the other hand, if you do

```
  mvps:?basename
& extract$(str$(!basename ".txt"),str$(!basename ".out"),\r\n,1417,(="0.0.0.0 "),(=" " ?|));
```

The first and second parameter are evaluated to <code>"mvps.txt"</code> and <code>"mvps.out"</code> and these values are passed to the extract function.

**Comment by xtradev** [comment]
Slightly intricate, so was not readily obvious to me. I overlooked the `!arg` part and the fact that `=` prevents the evaluation, but not unification (it does not prevent `?lineSuffix` from unifying with the argument).

And it's doable :-) Thank you for the detailed explanation. To the point and very helpful.
Here is the whole working code:

```
( extract
= inFileName, outFileName, lineSeparator, startOffset, linePrefix, lineSuffix, hosts
.
!arg:(?inFileName, ?outFileName, ?lineSeparator, ?startOffset, ?linePrefix, (=?lineSuffix))
& fil$(!inFileName,r)
& fil$(,STR,!lineSeparator)
& fil$(,SET,!startOffset)
&   whl
  ' ( fil$:(?line.?)
    & `(@(!line:!linePrefix ?host !lineSuffix)
       & !host \n !hosts:?hosts
       )
    )
& (fil$(,SET,-1)|)
& put$(str$!hosts,!outFileName,NEW)
);
extract$("mvps.txt","mvps.out",\r\n,1417,  "0.0.0.0 ",(=" " ?|));
extract$("sowc.txt","sowc.out",\r\n,3923,"127.0.0.1 ",(=(\t|" ") ?|));
)y
```

**Comment by BartJongejan** [comment]
Nice!

If speed matters and the input files contain many lines, you may want to "compile" the loop before evaluating it, so the variables <code>linePrefix</code> and <code>lineSuffix</code> are evaluated only once instead of each time a line has been read:

```
  &     
      ' ( whl
        ' ( fil$:(?line.?)
          & `( @(!line:()$linePrefix ?host ()$lineSuffix)
             & !host \n !hosts:?hosts
             )
          )
        )
    : (=?loop)
  & !loop
```

Here we have used macro expansion to hardwire the values of <code>linePrefix</code> and <code>lineSuffix</code> into the loop expression. The result of the macro expansion, without the <code>=</code> operator, is assigned to the variable <code>loop</code>. The iteration is performed by evaluating this variable once.

**Comment by xtradev** [comment]
Good tip, thank you. I achieved significant improvement with it.

The code was run over 6 hosts files, one of which is quite big:

```
came.txt   651212 bytes
hpho.txt 21945622 bytes 606290 lines (hpHosts file from hosts-file.net)
mvps.txt   510650 bytes
mwdl.txt    45522 bytes
sowc.txt   339907 bytes
yoyo.txt    64623 bytes
```

The timings are in milliseconds. The second call is made after the change:

```
q)\t \bracmat "get$\"extract.bra\""
28171
q)\t \bracmat "get$\"extract.bra\""
18421
```
